glean-core/python/glean/_process_dispatcher.py

```diff
@@ -0,0 +1,76 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This module implements a class to dispatch work to another process.
+This work always happens one-at-a-time.  That is, it doesn't create
+another worker process until the previous worker process has completed.
+
+This is used only by the PingUploadWorker (and the test-only feature to delete
+temporary data directories which has a potential race condition with the
+PingUploadWorker).
+"""
+
+import sys
+from typing import Optional, TYPE_CHECKING, Union
+
+
+if TYPE_CHECKING:
+    import multiprocessing  # noqa
+
+
+def _work_wrapper(func, args):
+    """
+    A wrapper to call a function and convert its boolean success value into a
+    process exitcode.
+    """
+    success = func(*args)
+
+    if success:
+        sys.exit(0)
+    else:
+        sys.exit(1)
+
+
+class ProcessDispatcher:
+    """
+    A class to dispatch work to another process. This work always happens
+    one-at-a-time. That is, it doesn't create another worker process until the
+    previous worker process has completed.
+
+    Since running a second process might block, this should only be used from
+    the worker thread in `_dispatcher.Dispatcher.
+    """
+
+    # Store the last run process object so we can `join` it when starting
+    # another process.
+    _last_process = None  # type: Optional[multiprocessing.Process]
+
+    @classmethod
+    def dispatch(cls, func, args) -> Union[bool, "multiprocessing.Process"]:
+        from . import Glean
+
+        if Glean._configuration._allow_multiprocessing:
+            # Only import the multiprocessing library if it's actually needed
+            import multiprocessing  # noqa
+
+            # We only want one of these processes running at a time, so if there's
+            # already one, join on it. PingUploadWorker.process is only triggered
+            # from a worker thread anyway, so this blocking will not block the main
```

Comment:
This is provably wrong: it's called in `atexit`.
So I think adjusting this slightly to say the same as the class doc: it might block, better run it off-the-main-thread

Reply:
That's a good point.  At least the `atexit` thing will only block in testing mode (when using temporary directories), so I'm not concerned about the implications of that.  But I'll fix the comment to be more clear.